### PR TITLE
cephadm: remove injected_args

### PR DESCRIPTION
**Pull request by sebastian-philipp**
This PR is based on #39141 and will contain those commits, unit it is merged

Avoid copying cephadm all the time.

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by sebastian-philipp** [comment]
replaces #37495

**Comment by liewegas** [comment]
jenkins test make check

File: src/cephadm/cephadm [hunk 1]
```diff
@@ -6601,11 +6577,9 @@ class CephadmDaemon():
...
     @property
     def binary_path(self):
-        return os.path.join(
-            self.ctx.data_dir,
-            self.fsid,
-            CephadmDaemon.bin_name
-        )
+        path = os.path.realpath(__file__)
+        assert os.path.isfile(path)
+        return path
```
Comment (sebastian-philipp) [inline src/cephadm/cephadm:6582]: @adk3798

**Comment by sebastian-philipp** [comment]
```
>       assert exporter.binary_path == "/var/lib/ceph/foobar/cephadm"
E       AssertionError: assert '/home/jenkin...phadm/cephadm' == '/var/lib/ceph/foobar/cephadm'
E         - /var/lib/ceph/foobar/cephadm
E         + /home/jenkins-build/build/workspace/ceph-pull-requests/src/cephadm/cephadm

**Comment by sebastian-philipp** [comment]
changelog: fix cephadm tox

File: src/pybind/mgr/cephadm/serve.py [hunk 2]
```diff
@@ -1063,16 +1067,29 @@ def _registry_login(self, host: str, url: Optional[str], username: Optional[str]
...
-        return code == 0
+            return self._deploy_cephadm_binary_conn(conn, host)
+
+    def _deploy_cephadm_binary_conn(self, conn: "BaseConnection", host: str) -> None:
+        _out, _err, code = remoto.process.check(
+            conn,
+            ['mkdir', '-p', f'/var/lib/ceph/{self.mgr._cluster_fsid}'])
+        if code:
+            msg = f"Unable to deploy the cephadm binary to {host}: {_err}"
+            self.log.warning(msg)
```
Comment (jmolmo) [inline src/pybind/mgr/cephadm/serve.py:1083]: maybe this kind of situations and the consequences deserves an "error" instead of  a "warning"

**Review by jmolmo — APPROVED** [review]
It works.!

File: src/pybind/mgr/cephadm/serve.py [hunk 3]
```diff
@@ -993,13 +986,24 @@ def _run_cephadm(self,
...
+                        stdin=stdin.encode('utf-8') if stdin is not None else None)
+                    if code == 2:
+                        out_ls, err_ls, code_ls = remoto.process.check(
+                            conn, ['ls', self.mgr.cephadm_binary_path])
+                        if code_ls == 2:
+                            self._deploy_cephadm_binary_conn(conn, host)
+                            out, err, code = remoto.process.check(
+                                conn,
+                                [python, self.mgr.cephadm_binary_path] + final_args,
+                                stdin=stdin.encode('utf-8') if stdin is not None else None)
```
Comment (mgfritch) [inline src/pybind/mgr/cephadm/serve.py:999]: (nit) the `remoto.process` logic and the logic above are duplicates ..  move into a helper func?

File: src/pybind/mgr/cephadm/serve.py [hunk 4]
```diff
@@ -993,13 +986,24 @@ def _run_cephadm(self,
...
                     out, err, code = remoto.process.check(
                         conn,
-                        [python, '-u'],
-                        stdin=script.encode('utf-8'))
+                        [python, self.mgr.cephadm_binary_path] + final_args,
+                        stdin=stdin.encode('utf-8') if stdin is not None else None)
+                    if code == 2:
+                        out_ls, err_ls, code_ls = remoto.process.check(
+                            conn, ['ls', self.mgr.cephadm_binary_path])
+                        if code_ls == 2:
```
Comment (mgfritch) [inline src/pybind/mgr/cephadm/serve.py:994]:
```suggestion
                        if code_ls != 0:
```
(nit) similar, but maybe any non-zero errcode?

Might also be good to extract this into a dedicated `cephadm_binary_exists` function?

File: src/pybind/mgr/cephadm/serve.py [hunk 5]
```diff
@@ -993,13 +986,24 @@ def _run_cephadm(self,
                 try:
                     out, err, code = remoto.process.check(
                         conn,
-                        [python, '-u'],
-                        stdin=script.encode('utf-8'))
+                        [python, self.mgr.cephadm_binary_path] + final_args,
+                        stdin=stdin.encode('utf-8') if stdin is not None else None)
+                    if code == 2:
```
Comment (mgfritch) [inline src/pybind/mgr/cephadm/serve.py:991]:
```suggestion
                    if code == errno.ENOENT:
```

(nit) I was somewhat confused why retval 2, but it's `file not found` from python.

**Review by mgfritch — APPROVED** [review]
lgtm! .. mostly nits around readability of the remoto connection handling, but the code is correct

**Comment by sebastian-philipp** [comment]
https://pulpito.ceph.com/swagner-2021-03-09_13:58:13-rados:cephadm-wip-swagner-testing-2021-03-09-1014-distro-basic-smithi/

* https://tracker.ceph.com/issues/49633
* dashboard e2e only runs on centos

**Comment by sebastian-philipp** [comment]
jenkins test make check

**Pull request merged**